### ISSUE TITLE
[ECS] [request]: Exec into all containers when pidMode=task

### ISSUE THREAD

**Issue by robertjli** [aws/containers-roadmap#2268]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**

https://repost.aws/questions/QUbsgWD4yERDuphXDKOWbNFA/ecs-execute-command-not-working-for-pid-mode-task

> At this time, the behavior of Amazon ECS is non-deterministic with respect to enableExecuteCommand when pidMode is set to task. The AWS SSM agent (which powers the feature) will be running in one of the containers only, but right now you cannot specify which container is the one in which it will run, nor can you specify that you want it to run in all of them.

It would be extremely useful to be able to exec into all containers for debugging.

**Which service(s) is this request for?**
Fargate, ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

Currently, when deploying a task with sidecar containers, it is impossible to debug issues in a given container unless it's specified as the first one (and from the accepted answer, that is not even a guarantee). Specifically, I'm running an agent container that collects process-level metrics, and the agent has subcommands that I would like to run to assist in debugging.

**Are you currently working around this issue?**

No current workaround.

**Additional context**
Anything else we should know?

**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

**Comment by kellenanker7** [comment]
this is a pretty serious gap IMO. it's not obvious and is also not called out anywhere in the ECS docs; it took me several hours of debugging and searching to find the AWS repost thread quoted in this bug report.

**Comment by jon94** [comment]
Yep. Experienced this too. There is no deterministic way to exec into a specific container when pidMode:task is defined on the task definition. Took me several hours of trial and error before identifying this pattern and finally found this issue here.

**Comment by suemayaheldursi** [comment]
Hello - this is a hug problem and preventing us from debugging issues with our sidecar setup for logging. Are there any workarounds or solutions???

**Comment by rfreuler** [comment]
same here... this is a rather big problem running security containers in a true sidecar setup we loose the ability to troubleshoot the containers. I'd love to see a fix for this...

**Comment by and-stuber** [comment]
This is impairing my Monitoring with Elasticsearch in production environment. Would be nice see it moving on soon.

**Comment by js-truework** [comment]
For our particular use case, it would also be acceptable if `pidMode` was a valid runtime option in boto3's `ECS.Client.run_task` or as an override. We don't ever really connect to existing running tasks (to avoid accidentally negatively impacting production web requests, etc), but rather spin up a task specifically for exec'ing into. So if our production containers could still have sidecar monitoring from datadog with pidMode set to task, but we have the option to spin up a new task based on a task definition but with pidMode being unset (or set to empty string or null or whatever is valid), that would be very helpful.

**Comment by sotzing** [comment]
Given the prevalence of sidecar container use - this seems like a major limitation/bug. 

We're using the Datadog sidecar (which requires use of pidMode=task for process monitoring) but not being able to Exec into the container (which we do for debugging purposes) is a major issue for us...

**Comment by albertorm95** [comment]
Any updates?

**Comment by bbrunodiad** [comment]
Any updates?

**Comment by ocha-heidi** [comment]
Any updates on this one? This is a big issue and we can't debug anything to find issues within our containers processes (we can ecs exec but it's painful)

**Comment by hussein-ayoub** [comment]
Any updates ?

**Comment by bbrunodiad** [comment]
Any updates ?

**Comment by arrudm** [comment]
Any updates?

**Comment by robertomoutinho** [comment]
@staceywhou, we would all love an update on this matter, as this is a significant blocker when using sidecar containers for monitoring or other activities.

**Comment by c-ameron** [comment]
so I guess the work around for this is to (unfortunately) install the aws-sam-cli in our application containers, correct?

**Comment by staceywhou** [comment]
Thanks everyone for surfacing the issue. The issue is a known limitation, where only one ECS Exec instance can be launched per PID namespace ([AWS Documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-exec.html#ecs-exec-considerations)). When sharing a PID namespace in a task, ECS Exec only allows you to connect to the container that starts last. A potential workaround is to create `dependsOn` relationships [API: Container Dependency](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html) to enforce that the container you want to ECS Exec into is the last one to start. We do not have an ETA, though we are currently looking into it on our end. We will share back with the thread once we have updates.

**Comment by nazar-mambetov** [comment]
@staceywhou 
> Thanks everyone for surfacing the issue. The issue is a known limitation, where only one ECS Exec instance can be launched per PID namespace ([AWS Documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-exec.html#ecs-exec-considerations)). When sharing a PID namespace in a task, ECS Exec only allows you to connect to the container that starts last. A potential workaround is to create `dependsOn` relationships [API: Container Dependency](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html) to enforce that the container you want to ECS Exec into is the last one to start. We do not have an ETA, though we are currently looking into it on our end. We will share back with the thread once we have updates.

I just tested the workaround by adding a condition to the app container, and it didn't work.
Still getting the same `TargetNotConnectedException` error.

The updated task definition:
```
      "dependsOn": [
        {
          "containerName": "datadog-agent",
          "condition": "START"
        },
        {
          "containerName": "fluent-bit",
          "condition": "START"
        }
      ],
```

**Comment by ron-sweet** [comment]
Any updates ?

**Comment by naorhaziz** [comment]
Any updates ?

**Comment by ron-sweet** [comment]
Any updates ?

**Comment by robertomoutinho** [comment]
> [@staceywhou](https://github.com/staceywhou)
> 
> > Thanks everyone for surfacing the issue. The issue is a known limitation, where only one ECS Exec instance can be launched per PID namespace ([AWS Documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-exec.html#ecs-exec-considerations)). When sharing a PID namespace in a task, ECS Exec only allows you to connect to the container that starts last. A potential workaround is to create `dependsOn` relationships [API: Container Dependency](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html) to enforce that the container you want to ECS Exec into is the last one to start. We do not have an ETA, though we are currently looking into it on our end. We will share back with the thread once we have updates.
> 
> I just tested the workaround by adding a condition to the app container, and it didn't work. Still getting the same `TargetNotConnectedException` error.
> 
> The updated task definition:
> 
> ```
>       "dependsOn": [
>         {
>           "containerName": "datadog-agent",
>           "condition": "START"
>         },
>         {
>           "containerName": "fluent-bit",
>           "condition": "START"
>         }
>       ],
> ```

Same here. The workaround did not worked.

@staceywhou Is there a chance that the `aws-guardduty-agent-r2GqQRp` sidecar can also "hijack" the ecs exec instance ?
I can't create a START dependency on that as it's name is created upon container startup (the hash at the end of the container name)

**Comment by JacksonDavenport** [comment]
For the workaround sake I found that using the condition of `HEALTHY` rather than `START` with those containers having some `healthCheck` seemed to allow it to consistently connect to the desired container. Granted you are still blocked from the other containers. 

For example gave access to the `app` container:
```json
{
    "containerDefinitions": [
        {
            "name": "app",
            "image": "example/app:latest",
            ...
            "dependsOn": [
                {
                    "containerName": "datadog-agent",
                    "condition": "HEALTHY"
                },
                {
                    "containerName": "log_router",
                    "condition": "HEALTHY"
                }
            ]
        },
        {
            "name": "datadog-agent",
            "image": "datadog/agent:7",
            ...
            "healthCheck": {
                "command": [
                    "CMD-SHELL",
                    "agent health"
                ],
                "interval": 30,
                "timeout": 5,
                "retries": 3,
                "startPeriod": 15
            }
        },
        {
            "name": "log_router",
            "image": "public.ecr.aws/aws-observability/aws-for-fluent-bit:stable",
            ...
            "healthCheck": {
                "command": [
                    "CMD-SHELL",
                    "curl -f http://127.0.0.1:2020/api/v1/uptime || exit 1"
                ],
                "interval": 10,
                "timeout": 5,
                "retries": 2,
                "startPeriod": 30
            },
            "firelensConfiguration": {
                "type": "fluentbit",
                "options": {
                    "config-file-type": "file",
                    "config-file-value": "/fluent-bit/configs/output-metrics-healthcheck.conf"
                }
            }
        }
    ]
}
```

For reference too this uses the fluent-bit health checks defined:
- https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/health-check
